cgp/genome.py

```diff
@@ -267,6 +267,181 @@ def randomize(self, rng: np.random.RandomState) -> None:
+            return
+
+        dna_for_reorder = self._dna.copy()
+
+        dependencies = self._determine_dependencies()
+
+        addable_nodes = set(int(idx_str) for idx_str, item in dependencies.items() if not item)
+
+        node_idx = self._n_inputs  # First position to be placed is after inputs
+        used_idx_list: List = []
+
+        while len(addable_nodes) > 0:
+            # pick one of the addable nodes
+            current_idx = rng.choice(list(addable_nodes))
+
+            dna_for_reorder = self._exchange_dna_segments(
+                dna_for_reorder, idx_old_location=current_idx, idx_new_location=node_idx
+            )
+
+            for _, current_dependencies in dependencies.items():
+                current_dependencies.discard(current_idx)
+
+            addable_nodes = self._update_addable_nodes(
+                dependencies, used_idx_list, idx_used=current_idx
+            )
+            node_idx += 1
+
+        self._update_gene_values(dna_for_reorder, used_idx_list)
+        self._replace_invalid_gene_values(dna_for_reorder, rng)
+
+        self.dna = dna_for_reorder
+
+    def _exchange_dna_segments(
+        self, dna_new: List[int], idx_old_location: int, idx_new_location: int
+    ) -> List[int]:
+        """ Moves a nodes dna from its old node location to a new location. """
+
+        node_dna = self._dna[
+            idx_old_location
+            * self._length_per_region : (idx_old_location + 1)
+            * self._length_per_region
+        ]
+        dna_new[
+            idx_new_location
+            * self._length_per_region : (idx_new_location + 1)
+            * self._length_per_region
+        ] = node_dna
+
+        return dna_new
+
+    def _update_addable_nodes(
+        self, dependencies: Dict, used_idx_list: List[int], idx_used: int
+    ) -> set:
+        """ Update the set of addable nodes,
+        by reevaluating which nodes haves no dependencies
+        and remove nodes which were already placed from the list.
+        """
+        addable_nodes = set(int(idx_str) for idx_str, item in dependencies.items() if not item)
+        used_idx_list.append(idx_used)
+        addable_nodes = addable_nodes.difference(used_idx_list)
+
+        return addable_nodes
+
+    def _update_gene_values(self, dna_for_reorder: List[int], used_idx_list: List[int]) -> None:
+        """ Update gene values to correspond to the new position of the corresponding node"""
+        for gene_idx, gene_value in enumerate(dna_for_reorder):
+            region_idx = gene_idx // self._length_per_region
+            if self._is_hidden_input_gene(gene_idx, region_idx) or self._is_output_input_gene(
+                gene_idx
+            ):
+                try:
+                    gene_value = self._n_inputs + used_idx_list.index(gene_value)
+                except ValueError:  # means the input is from an input node -> do nothing
+                    pass
```

Comment:
```suggestion
            if gene_value >= self._n_inputs:
                    gene_value = self._n_inputs + used_idx_list.index(gene_value)
```
isn't this clearer? if so, maybe we should have a helper function for this? `is_input_node`?